File: training/ruby/lib/de/sots/cellars_and_centaurs/adapter/persistence/repository.rb
# frozen_string_literal: true

require_relative '../../version'

module De::Sots::CellarsAndCentaurs::Adapter::Persistence
  class Repository
    include Enumerable

    def initialize
      @rows = {}
    end

    def save(row)
      @rows[key_of(row)] = row
    end

    def find_one(key)
      @rows[key]
    end

    def each(&block)
      @rows.values.each(&block)
    end

    private

    def key_of(row)
      raise NotImplementedError, "#{self.class} must implement key_of"
    end
  end
end

File: sample-projects/go/adapter/persistence/repository.go
package persistence

type Keyed interface {
	Key() string
}

type Repository[T Keyed] struct {
	rows map[string]T
}

func NewRepository[T Keyed]() Repository[T] {
	return Repository[T]{rows: make(map[string]T)}
}

func (r *Repository[T]) Save(row T) {
	r.rows[row.Key()] = row
}

func (r *Repository[T]) FindOne(key string) (T, bool) {
	row, ok := r.rows[key]
	return row, ok
}

func (r *Repository[T]) FindAll() []T {
	all := make([]T, 0, len(r.rows))
	for _, row := range r.rows {
		all = append(all, row)
	}
	return all
}
